### PR TITLE
Add CI job testing only Cabal with the latest GHC

### PR DESCRIPTION
**Pull request by andreabedini**
This is to build Cabal with boot packages form the latest GHC.

(cabal-install will include other dependencies which will likely exclude package versions release with newer GHCs).

Ping @Mikolaj @jneira @phadej 

---
Please include the following checklist in your PR:

* [ ] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#conventions).
* [ ] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog).
* [ ] The documentation has been updated, if necessary.

Please also shortly describe how you tested your change. Bonus points for added tests!

### CODE REVIEW

**Review by Mikolaj — APPROVED** [review]
That should help. Thank you.

File: .github/workflows/validate.yml [hunk 1]
```diff
@@ -46,6 +46,18 @@ jobs:
...
+          - os: "ubuntu-latest"
+            ghc: "9.2.1"
+            cli: "false"
+          - os: "macos-latest"
+            ghc: "9.2.1"
+            cli: "false"
+          - os: "windows-latest"
+            ghc: "9.2.1"
+            cli: "false"
+          #
```
Comment (jneira) [inline .github/workflows/validate.yml:60]: minor thing, an empty comment

**Comment by andreabedini** [comment]
I am not sure this is doing what I expected, I only see one `ubuntu-latest ghc-9.2.1` rather than two. I should have included the cli flag in the job title.

File: .github/workflows/validate.yml [hunk 2]
```diff
@@ -46,6 +46,18 @@ jobs:
           - os: "windows-latest"
             ghc: "8.4.4"
         include:
+          # this is to build Cabal with boot packages form the lastest GHC
+          # (cabal-install will include other dependencies which will likely exclude package versions release with newer GHCs)
+          - os: "ubuntu-latest"
+            ghc: "9.2.1"
```
[the remark this answers is not in the document]
Comment (andreabedini) [reply .github/workflows/validate.yml:52]: Yes but building Cabal and cabal-install together in a build plan will likely exclude new versions of the boot libraries through dependencies of cabal-install.

By the way. AFAIU elements of the build matrix don't have any specified semantic, so they should not get "collapsed" in any way. We had the tuple `(os: ubuntu-latest, ghc: 9.2.1, cli: null)` where `cli: null` gets interpreted as `true` in https://github.com/haskell/cabal/blob/a5309888fc0a33e511504c105907e8603a572143/.github/workflows/validate.yml#L109-L111 I am expecting this PR to add the tuple `(os: ubuntu-latest, ghc: 9.2.1, cli: false)`.

[the remark this answers is not in the document]
Comment (jneira) [reply .github/workflows/validate.yml:52]: aha, thanks for the explanation, could we add a comment about why we are building the same ghc twice?

File: .github/workflows/validate.yml [hunk 3]
```diff
@@ -46,6 +46,18 @@ jobs:
           - os: "windows-latest"
             ghc: "8.4.4"
         include:
+          # this is to build Cabal with boot packages form the lastest GHC
```
Comment (andreasabel) [inline .github/workflows/validate.yml:49]: lastest -> latest

File: .github/workflows/validate.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreasabel) [reply .github/workflows/validate.yml:52]: This feels like a hack to me.  
It would be clearer if `cli: [true]` was added to the build matrix, imo.

**Review by andreasabel — CHANGES_REQUESTED** [review]

File: .github/workflows/validate.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (andreabedini) [reply .github/workflows/validate.yml:52]: @andreasabel I agree. @jneira I hope you don't mind if I change that.

**Comment by andreabedini** [comment]
I am thinking we should try a different approach here.

Our "validate" workflow should tests whether or not changes we make to the repo break the assumptions defined in the test suites.

Testing Cabal against the latest versions of the boot libraries is a completely different thing and should belong to a separate workflow.

If there's no objections, I will make a separate workflow that would run on a schedule (once a week?) and also on demand, that only builds Cabal. Does that sound right? (Btw, should I use `--allow-newer`?).

**Comment by andreasabel** [comment]
> make a separate workflow that would run on a schedule (once a week?) and also on demand, that only builds Cabal. Does that sound right? 

That sounds very reasonable.

> (Btw, should I use `--allow-newer`?).

In my experience, a "all in" `--allow-newer` often leads to bad build plans, as very old versions of package might suddenly become candidates.  But using it selectively could be a useful tool.  (Maybe something like `--allow-newer='Cabal:*'` to restrict relaxation to `Cabal`'s dependency list only.)

File: .github/workflows/validate.yml (hunk 2 above)
[the remark this answers is not in the document]
Comment (jneira) [reply .github/workflows/validate.yml:52]: well, no other than being able to reverse the default behaviour only changing the condition iirc, but agree we could be more explicit here and use the variable directly to reverse it

**Comment by andreabedini** [comment]
> Maybe something like `--allow-newer='Cabal:*'` to restrict relaxation to `Cabal`'s dependency list only

Oh yes, sorry, I absolutely meant that. Thank you for correcting me.

**Comment by jneira** [comment]
> If there's no objections, I will make a separate workflow that would run on a schedule (once a week?) and also on demand, that only builds Cabal. Does that sound right? (Btw, should I use --allow-newer?).

But we should think in some way to make visible the fail, to fix it asap. Scheduled workflows fails are notified via github but maybe we would need another channel (email?)

**Comment by Mikolaj** [comment]
> but maybe we would need another channel (email?)

Sending failures to all maintainers, with a possibility of opt-out, sounds a good idea to me.

**Pull request closed without merging**